pandas/core/arrays/string_arrow.py

```diff
@@ -0,0 +1,619 @@
+            backfill / bfill: use NEXT valid observation to fill gap
+        limit : int, default None
+            If method is specified, this is the maximum number of consecutive
+            NaN values to forward/backward fill. In other words, if there is
+            a gap with more than this number of consecutive NaNs, it will only
+            be partially filled. If method is not specified, this is the
+            maximum number of entries along the entire axis where NaNs will be
+            filled.
+
+        Returns
+        -------
+        filled : ExtensionArray with NA/NaN filled
+        """
+
+        value, method = validate_fillna_kwargs(value, method)
+
+        mask = self.isna()
+
+        if is_array_like(value):
+            if len(value) != len(self):
+                raise ValueError(
+                    f"Length of 'value' does not match. Got ({len(value)}) "
+                    f"expected {len(self)}"
+                )
+            value = value[mask]
+
+        if mask.any():
+            if method is not None:
+                func = get_fill_func(method)
+                new_values = func(self.astype(object), limit=limit, mask=mask)
+                new_values = self._from_sequence(new_values)
+            else:
+                # fill with value
+                new_values = self.copy()
+                new_values[mask] = value
+        else:
+            new_values = self.copy()
+        return new_values
+
+    def _reduce(self, name, skipna=True, **kwargs):
+        if name in ["min", "max"]:
+            return getattr(self, name)(skipna=skipna)
+
+        raise TypeError(f"Cannot perform reduction '{name}' with string dtype")
+
+    @property
+    def nbytes(self) -> int:
+        """
+        The number of bytes needed to store this object in memory.
+        """
+        return self.data.nbytes
+
+    def isna(self) -> np.ndarray:
+        """
+        Boolean NumPy array indicating if each value is missing.
+
+        This should return a 1-D array the same length as 'self'.
+        """
+        # TODO: Implement .to_numpy for ChunkedArray
+        return self.data.is_null().to_pandas().values
+
+    def copy(self) -> ExtensionArray:
+        """
+        Return a copy of the array.
+
+        Parameters
+        ----------
+        deep : bool, default False
+            Also copy the underlying data backing this array.
+
+        Returns
+        -------
+        ExtensionArray
+        """
+        return type(self)(self.data)
+
+    def _cmp_method(self, other, op):
+        from pandas.arrays import BooleanArray
+
+        pc_func = ARROW_CMP_FUNCS[op.__name__]
+        if isinstance(other, (ABCSeries, ABCDataFrame, ABCIndex)):
+            return NotImplemented
+        if isinstance(other, ArrowStringArray):
+            result = pc_func(self.data, other.data)
+        elif is_scalar(other):
+            result = pc_func(self.data, pa.scalar(other))
```

Comment:
We might need to have some additional checks here for the scalars: 1) we should check for the scalar being a string, as otherwise pyarrow will raise an error about kernel not being implemented. If it is not a string, I think we can return an all-False array? 
2) also check for the scalar being a missing value (`pa.scalar(pd.NA)` doesn't work, you would need something like `pa.scalar(pd.NA, from_pandas=True)`, but even that should also be given a `pa.string()` type to ensure the proper type.

Reply:
> We might need to have some additional checks here for the scalars: 1) we should check for the scalar being a string, as otherwise pyarrow will raise an error about kernel not being implemented. If it is not a string, I think we can return an all-False array?

ok. we defer to pyarrow so should expected it to work, but could try/except ArrowNotImplementedError

> 2) also check for the scalar being a missing value (`pa.scalar(pd.NA)` doesn't work, you would need something like `pa.scalar(pd.NA, from_pandas=True)`, but even that should also be given a `pa.string()` type to ensure the proper type.

I think converting scalar to pa.scalar prior to dispatch makes sense. 

we have a test for this, test_comparison_methods_scalar which is failing for the reasons you state. was not originally planning to address the failures in pandas/core/arrays/string_arrow.py in this PR. but can do if you prefer.

Reply:
> but can do if you prefer.

maybe not so straightforward. pyarrow segfaults

```
>>> import pyarrow as pa
>>> import pyarrow.compute as pc
>>> import pandas as pd
>>>
>>>
>>> pa.__version__
'2.0.0'
>>>
>>> arr = pa.array(["a", "b", "c"], type=pa.string())
>>> arr
<pyarrow.lib.StringArray object at 0x000001C56CEFBB80>
[
  "a",
  "b",
  "c"
]
>>>
>>> scalar = pa.scalar(pd.NA, from_pandas=True, type=pa.string())
>>> scalar
<pyarrow.StringScalar: None>
>>>
>>> pc.equal(arr, scalar)
(pandas-dev)
```

Reply:
> > We might need to have some additional checks here for the scalars: 1) we should check for the scalar being a string, as otherwise pyarrow will raise an error about kernel not being implemented. If it is not a string, I think we can return an all-False array?
> 
> ok. we defer to pyarrow so should expected it to work, but could try/except ArrowNotImplementedError

this doesn't work for StringArray either and we have no tests for this case. probably need to fix and add test as precursor.

```
>>> import pandas as pd
>>>
>>> pd.__version__
'1.2.0.dev0+1249.g7bee5e2930'
>>>
>>> arr = pd.array(["a", None, "b"], dtype="string")
>>> arr
<StringArray>
['a', <NA>, 'b']
Length: 3, dtype: string
>>>
>>> arr > 42
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "c:\users\simon\pandas\pandas\core\ops\common.py", line 65, in new_method

    return method(self, other)
  File "c:\users\simon\pandas\pandas\core\arraylike.py", line 38, in __gt__
    return self._cmp_method(other, operator.gt)
  File "c:\users\simon\pandas\pandas\core\arrays\string_.py", line 362, in _cmp_
method
    result[valid] = op(self._ndarray[valid], other)
TypeError: '>' not supported between instances of 'str' and 'int'
>>>
```

Reply:
> maybe not so straightforward. pyarrow segfaults

Yeah, I noticed that yesterday as well, have opened an issue for that -> https://issues.apache.org/jira/browse/ARROW-10578

Now, that only happens for a missing value, so we can explicitly check for that case. 
I would handle that in this PR if that's OK for you.

Reply:
> this doesn't work for StringArray either and we have no tests for this case. probably need to fix and add test as precursor.

For that larger / smaller comparisons, you are right and that can be handled for both StringArray implementations afterwards in a searpate PR, but the basic equality works fine for StringArray:

```
In [10]: arr = pd.array(["a", None, "b"], dtype="string")
    ...: arr == 42
Out[10]: 
<BooleanArray>
[False, <NA>, False]
Length: 3, dtype: boolean
```

Reply:
ok. will investigate basic equality here. (and add tests with xfail for the others)

Reply:
> Now, that only happens for a missing value, so we can explicitly check for that case.
> I would handle that in this PR if that's OK for you.

makes sense

Reply:
have split test_comparison_methods_scalar and added a test for comparison against non-string type.

have added a fallback implementation rather than any pre-checks. We could maybe add a note to remove catching the ArrowInvalid and the isna check on other when creating the mask once https://issues.apache.org/jira/browse/ARROW-10578 is fixed or leave this as part of the fallback